Expose TCP connection establishment information

**Jetty version(s)**
10+

**Enhancement Description**
Would be nice to have `HttpClient` (or some other component such as `ClientConnector`) emit events for TCP connection establishment/failure (a begin event, and a success/fail event, for a given `SocketAddress`).

This could be useful to track destination servers that are faulty or take a long time to establish a connection.

Here is the stack trace of Throwable error:

```
java.util.concurrent.ExecutionException: java.net.ConnectException: Connection refused
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at org.eclipse.jetty.client.transport.HttpRequest.send(HttpRequest.java:707)
	at arsenal.jetty.httpclient.emit.Solution.main(Solution.java:45)
Caused by: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.Net.pollConnect(Native Method)
	at java.base/sun.nio.ch.Net.pollConnectNow(Net.java:672)
	at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:946)
	at org.eclipse.jetty.io.SelectorManager.doFinishConnect(SelectorManager.java:358)
	at org.eclipse.jetty.io.ManagedSelector.processConnect(ManagedSelector.java:346)
	at org.eclipse.jetty.io.ManagedSelector$SelectorProducer.processSelected(ManagedSelector.java:684)
	at org.eclipse.jetty.io.ManagedSelector$SelectorProducer.produce(ManagedSelector.java:539)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.produceTask(AdaptiveExecutionStrategy.java:512)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:258)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.produce(AdaptiveExecutionStrategy.java:195)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:979)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1209)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1164)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

--- a/documentation/jetty/modules/code/examples/src/main/java/org/eclipse/jetty/docs/programming/client/http/HTTPClientDocs.java
+++ b/documentation/jetty/modules/code/examples/src/main/java/org/eclipse/jetty/docs/programming/client/http/HTTPClientDocs.java
@@ -20,10 +20,13 @@ import java.io.OutputStream;
 import java.net.SocketAddress;
 import java.net.URI;
 import java.nio.ByteBuffer;
+import java.nio.channels.SocketChannel;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.List;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.TimeUnit;
 import javax.net.ssl.SSLEngine;
 import javax.net.ssl.SSLException;
@@ -1232,5 +1235,39 @@ public class HTTPClientDocs
             })
             .send();
         // end::connectionInformation[]
+    }
+
+    public void connectListener() throws Exception
+    {
+        // tag::connectListener[]
+        ClientConnector clientConnector = new ClientConnector();
+        clientConnector.addEventListener(new ClientConnector.ConnectListener()
+        {
+            private final ConcurrentMap<SocketChannel, Long> times = new ConcurrentHashMap<>();
+
+            @Override
+            public void onConnectBegin(SocketChannel socketChannel, SocketAddress socketAddress)
+            {
+                times.put(socketChannel, System.nanoTime());
+            }
+
+            @Override
+            public void onConnectSuccess(SocketChannel socketChannel)
+            {
+                Long begin = times.remove(socketChannel);
+                System.getLogger("connection").log(INFO, "established in %d ns", System.nanoTime() - begin);
+            }
+
+            @Override
+            public void onConnectFailure(SocketChannel socketChannel, SocketAddress socketAddress, Throwable failure)
+            {
+                Long begin = times.remove(socketChannel);
+                System.getLogger("connection").log(INFO, "failed in %d ns", System.nanoTime() - begin);
+            }
+        });
+
+        HttpClient httpClient = new HttpClient(new HttpClientTransportOverHTTP(clientConnector));
+        httpClient.start();
+        // end::connectListener[]
     }
 }

--- a/documentation/jetty/modules/programming-guide/pages/client/http.adoc
+++ b/documentation/jetty/modules/programming-guide/pages/client/http.adoc
@@ -541,7 +541,7 @@ include::code:example$src/main/java/org/eclipse/jetty/docs/programming/client/ht
 ----
 
 [[connection-information]]
-=== Request Connection Information
+=== Connection Information
 
 In order to send a request, it is necessary to obtain a connection, as explained in the xref:request-processing[request processing section].
 
@@ -561,6 +561,25 @@ The connection information is only available when the request is associated with
 This means that the connection is not available in the _request queued_ event, but only starting from the _request begin_ event.
 For more information about request events, see xref:non-blocking[this section].
 ====
+
+[[connection-events]]
+=== Connection Events
+
+In order to send HTTP requests, a connection is necessary, as explained in the xref:request-processing[request processing section].
+
+HTTP/1.1 and HTTP/2 use `SocketChannel.connect(SocketAddress)` to establish a connection with the server, either via the TCP transport or via the Unix-Domain transport.
+
+You can listen to these `connect` events using a `ClientConnector.ConnectListener`, for example to record connection establishment times:
+
+[,java,indent=0]
+----
+include::code:example$src/main/java/org/eclipse/jetty/docs/programming/client/http/HTTPClientDocs.java[tags=connectListener]
+----
+
+This could be particularly useful when you notice that your client application seem "slow" to send requests.
+The `connect begin` and `connect success` events, along with the `request queued` and `request begin` event (detailed xref:non-blocking[here]), allow you to understand whether it is the server being slow at accepting connections, or it is the client being slow at processing queued requests.
+
+Once the low-level connection has been established, you can be notified of connection events using a `Connection.Listener`, or more concretely `ConnectionStatistics`, as described in xref:arch/io.adoc#connection-listener[this section].
 
 [[configuration]]
 == HttpClient Configuration

--- a/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java
+++ b/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java
@@ -457,12 +457,11 @@ public class ClientConnector extends ContainerLifeCycle
             boolean connected = true;
             if (channel instanceof SocketChannel socketChannel)
             {
-                final SocketAddress socketAddress = address;
                 boolean blocking = isConnectBlocking() && address instanceof InetSocketAddress;
                 if (LOG.isDebugEnabled())
                     LOG.debug("Connecting {} to {}", blocking ? "blocking" : "non-blocking", address);
 
-                notifyConnectBegin(socketChannel, socketAddress);
+                notifyConnectBegin(socketChannel, address);
                 if (blocking)
                 {
                     socketChannel.socket().connect(address, (int)getConnectTimeout().toMillis());
@@ -473,6 +472,8 @@ public class ClientConnector extends ContainerLifeCycle
                 {
                     socketChannel.configureBlocking(false);
                     connected = socketChannel.connect(address);
+                    if (connected)
+                        notifyConnectSuccess(socketChannel);
                 }
             }
             else
@@ -494,7 +495,7 @@ public class ClientConnector extends ContainerLifeCycle
             if (x.getClass() == SocketException.class)
                 x = new SocketException("Could not connect to " + address).initCause(x);
             IO.close(channel);
-            connectFailed(channel, x, context);
+            connectFailed(channel, address, x, context);
         }
     }
 
@@ -580,14 +581,75 @@ public class ClientConnector extends ContainerLifeCycle
             promise.failed(failure);
     }
 
-    protected void connectFailed(SelectableChannel channel, Throwable failure, Map<String, Object> context)
+    protected void connectFailed(SelectableChannel channel, SocketAddress address, Throwable failure, Map<String, Object> context)
     {
         if (LOG.isDebugEnabled())
-            LOG.debug("Could not connect to {}", context.get(REMOTE_SOCKET_ADDRESS_CONTEXT_KEY));
-        notifyConnectFailure((SocketChannel)channel, failure);
+            LOG.debug("Could not connect to {}", address);
+        notifyConnectFailure((SocketChannel)channel, address, failure);
         Promise<?> promise = (Promise<?>)context.get(CONNECTION_PROMISE_CONTEXT_KEY);
         if (promise != null)
             promise.failed(failure);
+    }
+
+    @Override
+    public boolean addEventListener(EventListener listener)
+    {
+        if (listener instanceof ConnectListener connectListener)
+            return listeners.add(connectListener);
+        return super.addEventListener(listener);
+    }
+
+    @Override
+    public boolean removeEventListener(EventListener listener)
+    {
+        if (listener instanceof ConnectListener connectListener)
+            return listeners.remove(connectListener);
+        return super.removeEventListener(listener);
+    }
+
+    private void notifyConnectBegin(SocketChannel socketChannel, SocketAddress socketAddress)
+    {
+        for (ConnectListener listener : listeners)
+        {
+            try
+            {
+                listener.onConnectBegin(socketChannel, socketAddress);
+            }
+            catch (Throwable x)
+            {
+                LOG.info("failure notifying listener {}", listener, x);
+            }
+        }
+    }
+
+    private void notifyConnectSuccess(SocketChannel socketChannel)
+    {
+        for (ConnectListener listener : listeners)
+        {
+            try
+            {
+                listener.onConnectSuccess(socketChannel);
+            }
+            catch (Throwable x)
+            {
+                LOG.info("failure notifying listener {}", listener, x);
+            }
+        }
+    }
+
+    private void notifyConnectFailure(SocketChannel socketChannel, SocketAddress socketAddress, Throwable throwable)
+    {
+        for (ConnectListener listener : listeners)
+        {
+            try
+            {
+                listener.onConnectFailure(socketChannel, socketAddress, throwable);
+            }
+            catch (Throwable x)
+            {
+                LOG.info("failure notifying listener {}", listener, x);
+            }
+        }
     }
 
     protected class ClientSelectorManager extends SelectorManager
@@ -634,17 +696,20 @@ public class ClientConnector extends ContainerLifeCycle
         }
 
         @Override
-        protected void connectionFailed(SelectableChannel channel, Throwable failure, Object attachment)
+        public void connectionSucceeded(SelectableChannel channel)
         {
-            @SuppressWarnings("unchecked")
-            Map<String, Object> context = (Map<String, Object>)attachment;
-            connectFailed(channel, failure, context);
+            super.connectionSucceeded(channel);
+            notifyConnectSuccess((SocketChannel)channel);
         }
 
         @Override
-        public void connectSuccess(SelectableChannel channel)
+        protected void connectionFailed(SelectableChannel channel, Throwable failure, Object attachment)
         {
-            listeners.forEach(listener -> listener.onConnectSuccess((SocketChannel)channel));
+            super.connectionFailed(channel, failure, attachment);
+            @SuppressWarnings("unchecked")
+            Map<String, Object> context = (Map<String, Object>)attachment;
+            SocketAddress address = (SocketAddress)context.get(REMOTE_SOCKET_ADDRESS_CONTEXT_KEY);
+            connectFailed(channel, address, failure, context);
         }
     }
 
@@ -770,119 +835,44 @@ public class ClientConnector extends ContainerLifeCycle
     }
     
     /**
-     * <p>A listener emits events for TCP connection establishment and failure</p>
-     * {@link #ClientConnector.ConnectListener} is notified of the events:
+     * <p>A listener for events about {@link SocketChannel#connect(SocketAddress)}.</p>
+     * <p>The events are:</p>
      * <ul>
-     * <li>Beginning of connect attempt</li>
-     * <li>Success of connect attempt</li>
-     * <li>Failed connect attempt</li>
+     * <li>{@link ConnectListener#onConnectBegin(SocketChannel, SocketAddress) begin}, just before the {@code connect()} call</li>
+     * <li>{@link ConnectListener#onConnectSuccess(SocketChannel) success}, when the {@code connect()} call succeeds</li>
+     * <li>{@link ConnectListener#onConnectFailure(SocketChannel, SocketAddress, Throwable) failure}, when the {@code connect()} call fails</li>
      * </ul>
-     * As an example {@link #ClientConnector.ConnectListener} can be used to trace attempts of connect for HttpClient
      */
     public interface ConnectListener extends EventListener
     {
         /**
-         * <p>Callback method is invoked for the very beginning of connect attempt</p>
-         * <p>The {@code socketChannel} parameter can be used to extract socket channel information</p>
-         * <p>The {@code socketAddress} parameter can be used to extract socket address information</p>
-         * 
-         * @param socketChannel the socket channel which is registered with connect attempt
-         * @param socketAddress the socket address which is registered with connect attempt
+         * <p>Callback method invoked just before a {@link SocketChannel#connect(SocketAddress)} call.</p>
+         *
+         * @param socketChannel the local socket channel that is about to connect
+         * @param socketAddress the remote socket address to connect to
          */
-        public void onConnectBegin(SocketChannel socketChannel, SocketAddress socketAddress);
-        
-        /**
-         * <p>Callback method is invoked for successfully connect attempt</p>
-         * <p>The {@code socketChannel} parameter can be used to extract socket channel information</p>
-         * 
-         * @param socketChannel the socket channel which is registered with connect attempt
-         */
-        public void onConnectSuccess(SocketChannel socketChannel);
-        
-        /**
-         * <p>Callback method is invoked for failed connect attempt</p>
-         * <p>The {@code socketChannel} parameter can be used to extract socket channel information</p>
-         * <p>The {@code throwable} parameter can be used to extract exception information of connect attempt</p>
-         * 
-         * @param socketChannel the socket channel which is registered with connect attempt
-         */
-        public void onConnectFailure(SocketChannel socketChannel, Throwable throwable);
-    }
-
-    @Override
-    public boolean addEventListener(EventListener listener)
-    {
-        if (listener instanceof ConnectListener connectListener)
-            return listeners.add(connectListener);
-        return super.addEventListener(listener);
-    }
-
-    @Override
-    public boolean removeEventListener(EventListener listener)
-    {
-        if (listener instanceof ConnectListener connectListener)
-            return listeners.remove(connectListener);
-        return super.removeEventListener(listener);
-    }
-    
-    /**
-     * <p>A wrapper for ConnectListener.notifyConnectBegin() call</p>
-     * @param socketChannel {@link #SocketChannel} registered for connect attempt
-     * @param socketAddress {@link #SocketAddress} registered for connect attempt
-     */
-    private void notifyConnectBegin(SocketChannel socketChannel, SocketAddress socketAddress)
-    {
-
-        for (ConnectListener listener : listeners)
+        default void onConnectBegin(SocketChannel socketChannel, SocketAddress socketAddress)
         {
-            try
-            {
-                listener.onConnectBegin(socketChannel, socketAddress);
-            }
-            catch (Throwable x)
-            {
-                x.printStackTrace();
-            }
         }
-    }
-    
-    /**
-     * <p>Wrapper for ConnectListener.onConnectSuccess() call</p>
-     * @param socketChannel {@link #SocketChannel} for successfully connect attempt
-     */
-    private void notifyConnectSuccess(SocketChannel socketChannel)
-    {
 
-        for (ConnectListener listener : listeners)
-        {   
-            try
-            {
-                listener.onConnectSuccess(socketChannel);
-            }
-            catch (Throwable x)
-            {
-                x.printStackTrace();
-            }
-        }
-    }
-    
-    /**
-     * <p>Wrapper for ConnectListener.onConnectFailure() call</p>
-     * @param socketChannel {@link #SocketChannel} registered for failed connect attempt
-     * @param throwable {@link #Throwable} exception arisen with failed attempt
-     */
-    private void notifyConnectFailure(SocketChannel socketChannel, Throwable throwable)
-    {
-        for (ConnectListener listener : listeners)
+        /**
+         * <p>Callback method invoked when a {@link SocketChannel#connect(SocketAddress)} call completes successfully.</p>
+         *
+         * @param socketChannel the local socket channel that succeeded to connect to the remote socket address
+         */
+        default void onConnectSuccess(SocketChannel socketChannel)
         {
-            try
-            {
-                listener.onConnectFailure(socketChannel, throwable);
-            }
-            catch (Throwable x)
-            {
-                x.printStackTrace();
-            }
+        }
+
+        /**
+         * <p>Callback method invoked when a {@link SocketChannel#connect(SocketAddress)} call completes with a failure.</p>
+         *
+         * @param socketChannel the local socket channel that failed to connect to the remote socket address
+         * @param socketAddress the remote socket address to connect to
+         * @param failure the failure cause
+         */
+        default void onConnectFailure(SocketChannel socketChannel, SocketAddress socketAddress, Throwable failure)
+        {
         }
     }
 }

--- a/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java
+++ b/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java
@@ -28,8 +28,11 @@ import java.nio.channels.SelectionKey;
 import java.nio.channels.SocketChannel;
 import java.nio.file.Path;
 import java.time.Duration;
+import java.util.EventListener;
+import java.util.List;
 import java.util.Map;
 import java.util.Objects;
+import java.util.concurrent.CopyOnWriteArrayList;
 import java.util.concurrent.Executor;
 
 import org.eclipse.jetty.util.IO;
@@ -93,6 +96,7 @@ public class ClientConnector extends ContainerLifeCycle
     }
 
     private final Configurator configurator;
+    private final List<ConnectListener> listeners = new CopyOnWriteArrayList<>();
     private Executor executor;
     private Scheduler scheduler;
     private ByteBufferPool byteBufferPool;
@@ -453,12 +457,16 @@ public class ClientConnector extends ContainerLifeCycle
             boolean connected = true;
             if (channel instanceof SocketChannel socketChannel)
             {
+                final SocketAddress socketAddress = address;
                 boolean blocking = isConnectBlocking() && address instanceof InetSocketAddress;
                 if (LOG.isDebugEnabled())
                     LOG.debug("Connecting {} to {}", blocking ? "blocking" : "non-blocking", address);
+
+                notifyConnectBegin(socketChannel, socketAddress);
                 if (blocking)
                 {
                     socketChannel.socket().connect(address, (int)getConnectTimeout().toMillis());
+                    notifyConnectSuccess(socketChannel);
                     socketChannel.configureBlocking(false);
                 }
                 else
@@ -486,7 +494,7 @@ public class ClientConnector extends ContainerLifeCycle
             if (x.getClass() == SocketException.class)
                 x = new SocketException("Could not connect to " + address).initCause(x);
             IO.close(channel);
-            connectFailed(x, context);
+            connectFailed(channel, x, context);
         }
     }
 
@@ -572,10 +580,11 @@ public class ClientConnector extends ContainerLifeCycle
             promise.failed(failure);
     }
 
-    protected void connectFailed(Throwable failure, Map<String, Object> context)
+    protected void connectFailed(SelectableChannel channel, Throwable failure, Map<String, Object> context)
     {
         if (LOG.isDebugEnabled())
             LOG.debug("Could not connect to {}", context.get(REMOTE_SOCKET_ADDRESS_CONTEXT_KEY));
+        notifyConnectFailure((SocketChannel)channel, failure);
         Promise<?> promise = (Promise<?>)context.get(CONNECTION_PROMISE_CONTEXT_KEY);
         if (promise != null)
             promise.failed(failure);
@@ -629,7 +638,13 @@ public class ClientConnector extends ContainerLifeCycle
         {
             @SuppressWarnings("unchecked")
             Map<String, Object> context = (Map<String, Object>)attachment;
-            connectFailed(failure, context);
+            connectFailed(channel, failure, context);
+        }
+
+        @Override
+        public void connectSuccess(SelectableChannel channel)
+        {
+            listeners.forEach(listener -> listener.onConnectSuccess((SocketChannel)channel));
         }
     }
 
@@ -751,6 +766,123 @@ public class ClientConnector extends ContainerLifeCycle
                     return new ChannelWithAddress(socketChannel, socketAddress);
                 }
             };
+        }
+    }
+    
+    /**
+     * <p>A listener emits events for TCP connection establishment and failure</p>
+     * {@link #ClientConnector.ConnectListener} is notified of the events:
+     * <ul>
+     * <li>Beginning of connect attempt</li>
+     * <li>Success of connect attempt</li>
+     * <li>Failed connect attempt</li>
+     * </ul>
+     * As an example {@link #ClientConnector.ConnectListener} can be used to trace attempts of connect for HttpClient
+     */
+    public interface ConnectListener extends EventListener
+    {
+        /**
+         * <p>Callback method is invoked for the very beginning of connect attempt</p>
+         * <p>The {@code socketChannel} parameter can be used to extract socket channel information</p>
+         * <p>The {@code socketAddress} parameter can be used to extract socket address information</p>
+         * 
+         * @param socketChannel the socket channel which is registered with connect attempt
+         * @param socketAddress the socket address which is registered with connect attempt
+         */
+        public void onConnectBegin(SocketChannel socketChannel, SocketAddress socketAddress);
+        
+        /**
+         * <p>Callback method is invoked for successfully connect attempt</p>
+         * <p>The {@code socketChannel} parameter can be used to extract socket channel information</p>
+         * 
+         * @param socketChannel the socket channel which is registered with connect attempt
+         */
+        public void onConnectSuccess(SocketChannel socketChannel);
+        
+        /**
+         * <p>Callback method is invoked for failed connect attempt</p>
+         * <p>The {@code socketChannel} parameter can be used to extract socket channel information</p>
+         * <p>The {@code throwable} parameter can be used to extract exception information of connect attempt</p>
+         * 
+         * @param socketChannel the socket channel which is registered with connect attempt
+         */
+        public void onConnectFailure(SocketChannel socketChannel, Throwable throwable);
+    }
+
+    @Override
+    public boolean addEventListener(EventListener listener)
+    {
+        if (listener instanceof ConnectListener connectListener)
+            return listeners.add(connectListener);
+        return super.addEventListener(listener);
+    }
+
+    @Override
+    public boolean removeEventListener(EventListener listener)
+    {
+        if (listener instanceof ConnectListener connectListener)
+            return listeners.remove(connectListener);
+        return super.removeEventListener(listener);
+    }
+    
+    /**
+     * <p>A wrapper for ConnectListener.notifyConnectBegin() call</p>
+     * @param socketChannel {@link #SocketChannel} registered for connect attempt
+     * @param socketAddress {@link #SocketAddress} registered for connect attempt
+     */
+    private void notifyConnectBegin(SocketChannel socketChannel, SocketAddress socketAddress)
+    {
+
+        for (ConnectListener listener : listeners)
+        {
+            try
+            {
+                listener.onConnectBegin(socketChannel, socketAddress);
+            }
+            catch (Throwable x)
+            {
+                x.printStackTrace();
+            }
+        }
+    }
+    
+    /**
+     * <p>Wrapper for ConnectListener.onConnectSuccess() call</p>
+     * @param socketChannel {@link #SocketChannel} for successfully connect attempt
+     */
+    private void notifyConnectSuccess(SocketChannel socketChannel)
+    {
+
+        for (ConnectListener listener : listeners)
+        {   
+            try
+            {
+                listener.onConnectSuccess(socketChannel);
+            }
+            catch (Throwable x)
+            {
+                x.printStackTrace();
+            }
+        }
+    }
+    
+    /**
+     * <p>Wrapper for ConnectListener.onConnectFailure() call</p>
+     * @param socketChannel {@link #SocketChannel} registered for failed connect attempt
+     * @param throwable {@link #Throwable} exception arisen with failed attempt
+     */
+    private void notifyConnectFailure(SocketChannel socketChannel, Throwable throwable)
+    {
+        for (ConnectListener listener : listeners)
+        {
+            try
+            {
+                listener.onConnectFailure(socketChannel, throwable);
+            }
+            catch (Throwable x)
+            {
+                x.printStackTrace();
+            }
         }
     }
 }

--- a/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java
+++ b/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java
@@ -350,6 +350,7 @@ public class ManagedSelector extends ContainerLifeCycle implements Dumpable
             {
                 if (connect.timeout.cancel())
                 {
+                    _selectorManager.connectSuccess(channel);
                     key.interestOps(0);
                     execute(new CreateEndPoint(connect, key));
                 }

--- a/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java
+++ b/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java
@@ -350,7 +350,7 @@ public class ManagedSelector extends ContainerLifeCycle implements Dumpable
             {
                 if (connect.timeout.cancel())
                 {
-                    _selectorManager.connectSuccess(channel);
+                    _selectorManager.connectionSucceeded(channel);
                     key.interestOps(0);
                     execute(new CreateEndPoint(connect, key));
                 }

--- a/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/SelectorManager.java
+++ b/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/SelectorManager.java
@@ -480,6 +480,14 @@ public abstract class SelectorManager extends ContainerLifeCycle implements Dump
         }
     }
 
+    /**
+     * <p>Callback method invoked when a connect was completed.</p>
+     * @param channel the channel that attempted the connect
+     */
+    protected void connectSuccess(SelectableChannel channel)
+    {
+    }
+    
     public interface SelectorManagerListener extends EventListener
     {
     }

--- a/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/SelectorManager.java
+++ b/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/SelectorManager.java
@@ -369,8 +369,18 @@ public abstract class SelectorManager extends ContainerLifeCycle implements Dump
     }
 
     /**
-     * <p>Callback method invoked when a non-blocking connect cannot be completed.</p>
-     * <p>By default it just logs with level warning.</p>
+     * <p>Callback method invoked when a non-blocking connect was succeeded.</p>
+     *
+     * @param channel the channel that attempted the connect
+     */
+    protected void connectionSucceeded(SelectableChannel channel)
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("Connected {}", channel);
+    }
+
+    /**
+     * <p>Callback method invoked when a non-blocking connect failed.</p>
      *
      * @param channel the channel that attempted the connect
      * @param ex the exception that caused the connect to fail
@@ -378,7 +388,8 @@ public abstract class SelectorManager extends ContainerLifeCycle implements Dump
      */
     protected void connectionFailed(SelectableChannel channel, Throwable ex, Object attachment)
     {
-        LOG.warn(String.format("%s - %s", channel, attachment), ex);
+        if (LOG.isDebugEnabled())
+            LOG.debug("Failed to connect {}", channel, ex);
     }
 
     /**
@@ -480,14 +491,6 @@ public abstract class SelectorManager extends ContainerLifeCycle implements Dump
         }
     }
 
-    /**
-     * <p>Callback method invoked when a connect was completed.</p>
-     * @param channel the channel that attempted the connect
-     */
-    protected void connectSuccess(SelectableChannel channel)
-    {
-    }
-    
     public interface SelectorManagerListener extends EventListener
     {
     }

--- a/jetty-core/jetty-tests/jetty-test-client-transports/pom.xml
+++ b/jetty-core/jetty-tests/jetty-test-client-transports/pom.xml
@@ -29,6 +29,11 @@
     </dependency>
     <dependency>
       <groupId>org.eclipse.jetty</groupId>
+      <artifactId>jetty-io</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.eclipse.jetty</groupId>
       <artifactId>jetty-proxy</artifactId>
       <scope>test</scope>
     </dependency>

--- a/jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/AbstractTest.java
+++ b/jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/AbstractTest.java
@@ -13,8 +13,11 @@
 
 package org.eclipse.jetty.test.client.transport;
 
+import java.io.IOException;
 import java.lang.management.ManagementFactory;
 import java.lang.reflect.AnnotatedElement;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
 import java.net.URI;
 import java.nio.file.Files;
 import java.nio.file.Path;
@@ -91,6 +94,7 @@ public class AbstractTest
     protected ServerQuicConfiguration serverQuicConfig;
     protected Server server;
     protected AbstractConnector connector;
+    protected ClientConnector clientConnector;
     protected HttpClient client;
     protected ArrayByteBufferPool.Tracking serverBufferPool;
     protected ArrayByteBufferPool.Tracking clientBufferPool;
@@ -366,32 +370,28 @@ public class AbstractTest
 
     protected HttpClientTransport newHttpClientTransport(Transport transport) throws Exception
     {
+        clientConnector = new ClientConnector();
+        clientConnector.setSelectors(1);
         return switch (transport)
         {
             case HTTP, HTTPS ->
             {
-                ClientConnector clientConnector = new ClientConnector();
-                clientConnector.setSelectors(1);
                 clientConnector.setSslContextFactory(newSslContextFactoryClient());
                 yield new HttpClientTransportOverHTTP(clientConnector);
             }
             case H2C, H2 ->
             {
-                ClientConnector clientConnector = new ClientConnector();
-                clientConnector.setSelectors(1);
                 clientConnector.setSslContextFactory(newSslContextFactoryClient());
                 HTTP2Client http2Client = new HTTP2Client(clientConnector);
                 yield new HttpClientTransportOverHTTP2(http2Client);
             }
             case H3 ->
             {
-                ClientConnector clientConnector = new ClientConnector();
-                clientConnector.setSelectors(1);
                 SslContextFactory.Client sslClient = newSslContextFactoryClient();
                 HTTP3Client http3Client = new HTTP3Client(new ClientQuicConfiguration(sslClient, null), clientConnector);
                 yield new HttpClientTransportOverHTTP3(http3Client);
             }
-            case FCGI -> new HttpClientTransportOverFCGI(1, "");
+            case FCGI -> new HttpClientTransportOverFCGI(clientConnector, "");
         };
     }
 
@@ -432,6 +432,16 @@ public class AbstractTest
         {
             if (connector instanceof QuicServerConnector)
                 ((QuicServerConnector)connector).getQuicConfiguration().setMaxBidirectionalRemoteStreams(maxRequestsPerConnection);
+        }
+    }
+
+    public static int freePort() throws IOException
+    {
+        try (ServerSocket server = new ServerSocket())
+        {
+            server.setReuseAddress(true);
+            server.bind(new InetSocketAddress("localhost", 0));
+            return server.getLocalPort();
         }
     }
 

--- a/jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/ConnectListenerTest.java
+++ b/jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/ConnectListenerTest.java
@@ -1,0 +1,109 @@
+package org.eclipse.jetty.test.client.transport;
+
+import java.net.ConnectException;
+import java.net.SocketAddress;
+import java.net.URI;
+import java.nio.channels.SocketChannel;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.Request;
+import org.eclipse.jetty.client.transport.HttpClientTransportOverHTTP;
+import org.eclipse.jetty.io.ClientConnector;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.ServerConnector;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class ConnectListenerTest
+{
+    static Server server;
+    static ServerConnector serverConnector;
+    static ClientConnector clientConnector;
+    static CountDownLatch beginLatch = new CountDownLatch(1);
+    static CountDownLatch succesfulLatch = new CountDownLatch(1);
+    static CountDownLatch failedLatch = new CountDownLatch(1);
+    
+    @BeforeAll
+    public static void start() throws Exception
+    {
+        server = new Server();
+        serverConnector = new ServerConnector(server);
+        server.addConnector(serverConnector);
+        server.start();
+        
+        clientConnector = new ClientConnector();
+        clientConnector.addEventListener(new ClientConnector.ConnectListener()
+        {
+            @Override
+            public void onConnectBegin(SocketChannel s, SocketAddress a)
+            {
+                beginLatch.countDown();
+            }
+
+            @Override
+            public void onConnectSuccess(SocketChannel s)
+            {
+                succesfulLatch.countDown();
+            }
+
+            @Override
+            public void onConnectFailure(SocketChannel s, Throwable x)
+            {
+                failedLatch.countDown();
+            }
+        });
+    }
+    
+    @AfterAll
+    public static void stop() throws Exception
+    {
+        server.stop();
+    }
+    
+    @Test
+    public void testConnectListenerSuccessful() throws Exception
+    {
+        try (HttpClient client = new HttpClient(new HttpClientTransportOverHTTP(clientConnector)))
+        {
+            client.start();
+            URI uri = new URI("http://localhost:" + serverConnector.getLocalPort());
+            Request req = client.newRequest(uri);
+            req.send();
+            
+            assertTrue(beginLatch.await(5, TimeUnit.SECONDS));
+            assertTrue(succesfulLatch.await(5, TimeUnit.SECONDS));
+            
+            client.stop();
+        }
+    }
+
+    @Test
+    public void testConnectListenerFailed() throws Exception
+    {
+        try (HttpClient client = new HttpClient(new HttpClientTransportOverHTTP(clientConnector)))
+        {
+            client.start();
+            URI uri = new URI("http://localhost:" + (serverConnector.getLocalPort() - 1));
+            Request req = client.newRequest(uri);
+            try
+            {
+                req.send();
+            }
+            catch (Exception e)
+            {
+                if (e instanceof ConnectException)
+                {
+                    assertTrue(beginLatch.await(5, TimeUnit.SECONDS));
+                    assertTrue(failedLatch.await(5, TimeUnit.SECONDS));
+                }
+            }
+               
+            client.stop();
+        }
+    }
+}

--- a/jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/ConnectListenerTest.java
+++ b/jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/ConnectListenerTest.java
@@ -1,3 +1,16 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
 package org.eclipse.jetty.test.client.transport;
 
 import java.net.ConnectException;
@@ -5,105 +18,130 @@ import java.net.SocketAddress;
 import java.net.URI;
 import java.nio.channels.SocketChannel;
 import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
 
-import org.eclipse.jetty.client.HttpClient;
-import org.eclipse.jetty.client.Request;
-import org.eclipse.jetty.client.transport.HttpClientTransportOverHTTP;
+import org.eclipse.jetty.client.ContentResponse;
+import org.eclipse.jetty.http.HttpStatus;
 import org.eclipse.jetty.io.ClientConnector;
-import org.eclipse.jetty.server.Server;
-import org.eclipse.jetty.server.ServerConnector;
-import org.junit.jupiter.api.AfterAll;
-import org.junit.jupiter.api.BeforeAll;
-import org.junit.jupiter.api.Test;
+import org.hamcrest.Matchers;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
 
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
-public class ConnectListenerTest
+public class ConnectListenerTest extends AbstractTest
 {
-    static Server server;
-    static ServerConnector serverConnector;
-    static ClientConnector clientConnector;
-    static CountDownLatch beginLatch = new CountDownLatch(1);
-    static CountDownLatch succesfulLatch = new CountDownLatch(1);
-    static CountDownLatch failedLatch = new CountDownLatch(1);
-    
-    @BeforeAll
-    public static void start() throws Exception
+    @ParameterizedTest
+    @MethodSource("transportsTCP")
+    public void testBeginSuccessBlocking(Transport transport) throws Exception
     {
-        server = new Server();
-        serverConnector = new ServerConnector(server);
-        server.addConnector(serverConnector);
-        server.start();
-        
-        clientConnector = new ClientConnector();
+        testBeginSuccess(transport, true);
+    }
+
+    @ParameterizedTest
+    @MethodSource("transportsTCP")
+    public void testBeginSuccessNonBlocking(Transport transport) throws Exception
+    {
+        testBeginSuccess(transport, false);
+    }
+
+    private void testBeginSuccess(Transport transport, boolean blocking) throws Exception
+    {
+        startServer(transport, new EmptyServerHandler());
+        prepareClient(transport);
+        client.setConnectBlocking(blocking);
+        CountDownLatch beginLatch = new CountDownLatch(1);
+        CountDownLatch successLatch = new CountDownLatch(1);
+        CountDownLatch failureLatch = new CountDownLatch(1);
         clientConnector.addEventListener(new ClientConnector.ConnectListener()
         {
             @Override
-            public void onConnectBegin(SocketChannel s, SocketAddress a)
+            public void onConnectBegin(SocketChannel socketChannel, SocketAddress socketAddress)
             {
                 beginLatch.countDown();
             }
 
             @Override
-            public void onConnectSuccess(SocketChannel s)
+            public void onConnectSuccess(SocketChannel socketChannel)
             {
-                succesfulLatch.countDown();
+                successLatch.countDown();
             }
 
             @Override
-            public void onConnectFailure(SocketChannel s, Throwable x)
+            public void onConnectFailure(SocketChannel socketChannel, SocketAddress socketAddress, Throwable failure)
             {
-                failedLatch.countDown();
+                failureLatch.countDown();
             }
         });
-    }
-    
-    @AfterAll
-    public static void stop() throws Exception
-    {
-        server.stop();
-    }
-    
-    @Test
-    public void testConnectListenerSuccessful() throws Exception
-    {
-        try (HttpClient client = new HttpClient(new HttpClientTransportOverHTTP(clientConnector)))
-        {
-            client.start();
-            URI uri = new URI("http://localhost:" + serverConnector.getLocalPort());
-            Request req = client.newRequest(uri);
-            req.send();
-            
-            assertTrue(beginLatch.await(5, TimeUnit.SECONDS));
-            assertTrue(succesfulLatch.await(5, TimeUnit.SECONDS));
-            
-            client.stop();
-        }
+        client.start();
+
+        ContentResponse response = client.newRequest(newURI(transport))
+            .timeout(5, TimeUnit.SECONDS)
+            .send();
+
+        assertTrue(beginLatch.await(5, TimeUnit.SECONDS));
+        assertTrue(successLatch.await(5, TimeUnit.SECONDS));
+        assertEquals(1, failureLatch.getCount());
+        assertEquals(HttpStatus.OK_200, response.getStatus());
     }
 
-    @Test
-    public void testConnectListenerFailed() throws Exception
+    @ParameterizedTest
+    @MethodSource("transportsTCP")
+    public void testBeginFailureBlocking(Transport transport) throws Exception
     {
-        try (HttpClient client = new HttpClient(new HttpClientTransportOverHTTP(clientConnector)))
+        testBeginFailure(transport, true);
+    }
+
+    @ParameterizedTest
+    @MethodSource("transportsTCP")
+    public void testBeginFailureNonBlocking(Transport transport) throws Exception
+    {
+        testBeginFailure(transport, false);
+    }
+
+    public void testBeginFailure(Transport transport, boolean blocking) throws Exception
+    {
+        startServer(transport, new EmptyServerHandler());
+        prepareClient(transport);
+        client.setConnectBlocking(blocking);
+        CountDownLatch beginLatch = new CountDownLatch(1);
+        CountDownLatch successLatch = new CountDownLatch(1);
+        CountDownLatch failureLatch = new CountDownLatch(1);
+        clientConnector.addEventListener(new ClientConnector.ConnectListener()
         {
-            client.start();
-            URI uri = new URI("http://localhost:" + (serverConnector.getLocalPort() - 1));
-            Request req = client.newRequest(uri);
-            try
+            @Override
+            public void onConnectBegin(SocketChannel socketChannel, SocketAddress socketAddress)
             {
-                req.send();
+                beginLatch.countDown();
             }
-            catch (Exception e)
+
+            @Override
+            public void onConnectSuccess(SocketChannel socketChannel)
             {
-                if (e instanceof ConnectException)
-                {
-                    assertTrue(beginLatch.await(5, TimeUnit.SECONDS));
-                    assertTrue(failedLatch.await(5, TimeUnit.SECONDS));
-                }
+                successLatch.countDown();
             }
-               
-            client.stop();
-        }
+
+            @Override
+            public void onConnectFailure(SocketChannel socketChannel, SocketAddress socketAddress, Throwable failure)
+            {
+                failureLatch.countDown();
+            }
+        });
+        client.start();
+
+        URI uri = newURI(transport);
+        // Use a different port to fail the TCP connect.
+        URI badURI = new URI(uri.getScheme(), uri.getUserInfo(), uri.getHost(), freePort(), uri.getPath(), uri.getQuery(), uri.getFragment());
+        ExecutionException failure = assertThrows(ExecutionException.class, () -> client.newRequest(badURI)
+            .timeout(5, TimeUnit.SECONDS)
+            .send());
+        assertTrue(beginLatch.await(5, TimeUnit.SECONDS));
+        assertTrue(failureLatch.await(5, TimeUnit.SECONDS));
+        assertEquals(1, successLatch.getCount());
+        assertThat(failure.getCause(), Matchers.instanceOf(ConnectException.class));
     }
 }

--- a/jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/HTTPDynamicTransportTest.java
+++ b/jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/HTTPDynamicTransportTest.java
@@ -13,9 +13,6 @@
 
 package org.eclipse.jetty.test.client.transport;
 
-import java.io.IOException;
-import java.net.InetSocketAddress;
-import java.net.ServerSocket;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.List;
@@ -74,6 +71,7 @@ import org.eclipse.jetty.util.thread.QueuedThreadPool;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
+import static org.eclipse.jetty.test.client.transport.AbstractTest.freePort;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.is;
@@ -516,15 +514,5 @@ public class HTTPDynamicTransportTest extends AbstractTransportTest
             .send();
 
         assertThat(response.getStatus(), is(HttpStatus.OK_200));
-    }
-
-    private static int freePort() throws IOException
-    {
-        try (ServerSocket server = new ServerSocket())
-        {
-            server.setReuseAddress(true);
-            server.bind(new InetSocketAddress("localhost", 0));
-            return server.getLocalPort();
-        }
     }
 }

